### DISCUSSION TITLE
gha-runner-scale-set-0.10.0

### DISCUSSION THREAD

**Discussion by Link-** [actions/actions-runner-controller#3850] [Announcements]
This release includes major improvements to the runner provisioning duration. In short, you should see less latency between queueing a workflow run and having a runner available to execute the job.

Make sure to check [#3832](https://github.com/actions/actions-runner-controller/pull/3832) and [#3848](https://github.com/actions/actions-runner-controller/pull/3848) for details on how to fine-tune that behavior.

### Major changes

1. Add exponential backoff when generating runner reg tokens [#3724](https://github.com/actions/actions-runner-controller/pull/3724)
1. Make EphemeralRunnerController MaxConcurrentReconciles configurable [#3832](https://github.com/actions/actions-runner-controller/pull/3832)
1. Make EphemeralRunnerReconciler create runner pods earlier [#3831](https://github.com/actions/actions-runner-controller/pull/3831)
1. Make k8s client rate limiter parameters configurable [#3848](https://github.com/actions/actions-runner-controller/pull/3848)

### Minor changes

1. Bump github.com/bradleyfalzon/ghinstallation/v2 from `2.8.0` to `2.12.0` [#3837](https://github.com/actions/actions-runner-controller/pull/3837)
2. Bump golang.org/x/crypto from `0.22.0` to `0.31.0` [#3844](https://github.com/actions/actions-runner-controller/pull/3844)
1. Update docs with details for the dashboard visualizations [#3696](https://github.com/actions/actions-runner-controller/pull/3696)

**Full Changelog**: https://github.com/actions/actions-runner-controller/compare/gha-runner-scale-set-0.9.3...gha-runner-scale-set-0.10.0

<hr /><em>This discussion was created from the release <a href='https://github.com/actions/actions-runner-controller/releases/tag/gha-runner-scale-set-0.10.0'>gha-runner-scale-set-0.10.0</a>.</em>

**Comment by Link-** [comment]
## TLDR; 

Thanks to contributions from @mumoshu @tfujiwar @int128 @ventsislav-georgiev and others, we shipped a couple of changes that will improve the duration between a workflow job queue and a runner being assigned to handle the job. This will also allow ARC to handle a higher amount of concurrent job runs.

> [!WARNING]
> The following requires advanced k8s knowledge and experience. Do not deploy these changes in a shared cluster. Always test & validate changes in a non-production environment. Follow [the guidelines for upgrading ARC](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/deploying-runner-scale-sets-with-actions-runner-controller#upgrading-arc) if you wish to enable these changes.

### How to configure the MaxConcurrentReconciles and k8s client rate limiters

3 new flags have been introduced to the [gha-runner-scale-set-controller](https://github.com/actions/actions-runner-controller/tree/master/charts/gha-runner-scale-set-controller) helm chart:

1. `runnerMaxConcurrentReconciles`: 
https://github.com/actions/actions-runner-controller/blob/59cb1d2c8b3c0076e79cc368767e3cc304c1bfc2/charts/gha-runner-scale-set-controller/values.yaml#L112
1. `k8sClientRateLimiterQPS`: 
https://github.com/actions/actions-runner-controller/blob/59cb1d2c8b3c0076e79cc368767e3cc304c1bfc2/charts/gha-runner-scale-set-controller/values.yaml#L140
1. `k8sClientRateLimiterBurst`: 
https://github.com/actions/actions-runner-controller/blob/59cb1d2c8b3c0076e79cc368767e3cc304c1bfc2/charts/gha-runner-scale-set-controller/values.yaml#L141

#### `runnerMaxConcurrentReconciles`

This will update the `MaxConcurrentReconciles` value for the ephemeral runner controller which will increase the number of go-routines processing the reconcile queue. The previous default value was `1` and now it's set to `2`.

**Keep in mind:**
1. If you observe rate limiting events from the Kubernetes API or GitHub's APIs, you might want to revert this value to `1`.
2. Higher values will also lead to higher CPU (and other resource consumption), so increase this number gradually and test the outcomes.
3. Test these changes during peak usage.

#### `k8sClientRateLimiterQPS` and `k8sClientRateLimiterBurst`

- `k8sClientRateLimiterQPS` (Queries Per Second): This parameter sets the maximum number of requests per second that the client can make to the Kubernetes API server. The default value is 20 QPS.
- `k8sClientRateLimiterBurst`: This parameter defines the maximum number of requests that can be made in a short burst. The default value is 30.

Depending on the configuration of your cluster, modifying these values **could** help further improve the overall ARC performance as it allows the controllers to make more requests to the k8s API server.

Modifying these values will not affect the limits imposed by the k8s API server! These are client side parameters.

Read more: https://kubernetes.io/docs/concepts/cluster-administration/flow-control/

**Comment by tfujiwar** [comment]
@Link- Hi. Thank you for including my https://github.com/actions/actions-runner-controller/pull/3848 in the release.

> - k8sClientRateLimiterQPS (Queries Per Second): This parameter sets the maximum number of requests per second that the client can make to the Kubernetes API server. The default value is 5 QPS
> - k8sClientRateLimiterBurst: This parameter defines the maximum number of requests that can be made in a short burst. The default value is 10

I think these are the default values of client-go but controller-runtime has another default values (20 QPS, 30 burst). I chose the latter as the default values of this controller.
https://github.com/actions/actions-runner-controller/pull/3833#discussion_r1865803728

**Reply by Link-** [reply]
I'll update my message, good callout. I think in the next release, we should lower these to the defaults (5 and 10). Some folks might already be running hot on their cluster and this might be a problematic change for them.

Revert PR: https://github.com/actions/actions-runner-controller/pull/3853

**Reply by tfujiwar** [reply]
No, you don't need to revert them. Those parameters were already 20 QPS and 30 bursts before 0.10.0 because the main function calls [`ctrl.GetConfigOrDie()`](https://github.com/actions/actions-runner-controller/blob/80d848339e5eeaa6b2cda3c4a5393dfcb4614794/main.go#L217) to get the config, which returns [20 QPS and 30 bursts](https://github.com/kubernetes-sigs/controller-runtime/blob/aea2e32a936584b06ae6f7992f856fe7292b0297/pkg/client/config/config.go#L101-L106).
